### ISSUE TITLE
Descriptor for instrument properties.

### ISSUE THREAD

**Issue by MatthieuDartiailh** [LabPy/labpy-discussion#5]
I think this is the most viable solutions (and is requested for the current implementation of the subsytems). I will describe what I chose to do in Eapii.
In Eapii each get/set operation is divided in three parts and each part can be customized separately using specially named method (to reserve to the case where the behavior is not standard, ie not well described by simply passing arguments to the descriptor) : 
- pre_get : used to perform some check before attempting to read the value. This can be for example checking that the current instrument mode makes sense for the requested value.
- get : This is dedicated to the communication. It can be handled automatically using the HasIProperty to which belongs the descriptor default_get_iproperty method. The get argument of the descriptor is passed to this method (it can be for example the message to send to the instrument).
- post_get : This is dedicated to converting the value to something nice for the user. For example is we do a simple pyvisa query we get a string, here we can convert it to an integer or use a dictionary to  map it a user friendly value.
- pre_set : Once again it can be used to perform checks but it can also convert the value so that it can be understood by the instrument. (we can do mapping, unit conversion, range checking, etc)
- set : Here also we can simply rely on the HasIProperty default method, which may have been customised for example to automatically add the channel id to the message before calling the driver method.
- post_set : Here we can check whether or not the operation succeeded (checking an error bit for example), by default this rely once more on a default method on the HasIProperty object.

Contrary to what was done in Lantz, I chose to subclass IProperties as to me it made for cleaner code. Another capabilities built-in was the possibility to automatically retry the operation if some error (specified in the driver) happened. This was requested by one of my users as some instrument would crash on a communication but work after reopening the connection.

I think there is room to improve this system so please comment !

**Comment by hgrecco** [comment]
I think there are two very related but distinct aspects here.

(1) The first is how to write properties with pre_ and post_ actions. 
(2) The syntax to create typical properties, i.e. the ones in which the user will not need to customize the pre and post as their are provided by the toolkit.

For (1), Lantz has the concept of processors which is very clean and testable convertors. This processors can be chained. Each Lantz Feat create a suitable chain of processors depending on the keyword arguments given to the Feat. Users can create their own processors as well. But I would like to argue against them because writing custom processors is something that the users do not do because they find it complicated. So the `pre_` and `post_` sounds good to me as they are a clear and proven methodology (for example used in Traits)

(2) Independent of 1, from the user side there at least two ways to add properties that with defined pre and post conditioning of values. The first, which I will call kwarg based, uses a single class and keyword arguments given to the constructor defines how they will act. This is how Lantz does it. The second which (subclass-based) implements the different behaviours in subclasses. This is how Eapii and PyVISA does it. Having implemented and used both, I would go for the subclass-based.

But I hope that for most cases users would not have to write the `pre_` nor the `post_`. And maybe we are able to avoid making them write the set and get as well. 

For example:

``` python
frequency = UnitFeature('FREQ?', 'The frequency is {:.3f}', 'FREQ {.3f}', 'volt')
```

meaning:

``` python
UnitFeature(<GET COMMAND>, <GET OUTPUT FORMAT>, <SET FORMAT>, <unit>)
```

**Comment by MatthieuDartiailh** [comment]
Yeah actually reaching what you suggest in your example is definitely what I am after too. I had no notion of output format in Eapii but it makes sense (using your stringparser ?) If everybody agrees as soon as we reach something that seems to fit I will be happy to propose a first implementation (in a bare repo).

**Comment by MatthieuDartiailh** [comment]
We also need to decide on a name for the base class. Feature like in Lantz ?

**Comment by arsenovic** [comment]
as for the name, i think Feature is good.

**Comment by MatthieuDartiailh** [comment]
Some raw ideas about what we may need (in term of kwargs) : 
- Feature : 
  - get, set, and 'extract' (for the potential get return message, this would be optional as if I ask for a float and get a string which can be evaluated I don't need the fancy extraction)
  - discard : list of cache entries which should be discarded when this value is set (this point was mentioned somewhere else by @alexforencich and I think it does make sense)
  - checks : list of string used to build some assertions (used for example to check the state of the instrument allow to set a particular value, ex {mode} == 'Voltage' to assert that the generator is not operating as a current source)
- Subclasses : 
  - Mapping based : This is devoted to cases where we need to map the some user friendly values to some instrument meaningful values. I can see three possible classes :
    - Mapping : Base class to which the mapping dict (user : instr) is given and the reverse dict is built internally.
    - Bool : For boolean values here the mapping map True and False to instrument values and we can have an aliases keyword to allow things like `output = 'ON'` meaning True. In that case the return value can however be surprising but I still think it worth it.
    - Register : Specific for 8-bit register which are a quite common things in instrument this would return a dict with the state of the bit.
  - Scalar based : Int, Float, Unicode for those we could have : 
    -values : set of allowed values
    - range : range of allowed values as either min, max, (step) or a name of a range attribute (with a getter name _range_... for example as the mapping between a range setting and an actual validator is most of the time not direct.
    - unit : for float only (unit specified by its name or by the name of the attribute storing it)

I am not being very imaginative here, simply repeating what Eapii looks like with some additions.
Please comment @LabPy/owners ! If we could just agree about Feature for the time being it would be nice so that my proposed implementation looks a bit real (still planned for next week-end as I will have some time then)

**Comment by MatthieuDartiailh** [comment]
Actually I forgot a bunch of things. Namely:
- logging support : should we log every communication. Honestly I find that a bit heavy, but perhaps we could have a debug mode in which the Feature are built with that capability.
- repeating operation after communication failure (unexpected timeout for example). Eapii does that by capturing the errors and checking them against the set of errors which should trigger a new attempt (defined on the driver).
- error checking : I think it important to make sure that all command have the expected result. One cheap way to do that is to read a status byte (or the answer of the instrument as some always answer). In Eapii this is done in post_get and post_set. Do we keep the idea ?

**Comment by hgrecco** [comment]
In general I am fine. A few comments about the Subclasses
- Bool: I am not sure about the aliases. One of the main goals should be the ability to exchange transparently instruments of the same kind. For example, if we start adding aliases depending on the manufacturer, exchanging devices will require changes to the users code. My suggestion: let's leave it out for now. We can always add it later if needed, but removing it is more difficult.
- Scalar: In lantz, we use `limits` instead of `range`. This actually came from the users during an early testing phase. The argument (and I agree) is that range collides with Python builtin. But more importantly the semantics are different as Python's `range` does not include the last point but usually we want that.

In addition, there are two behaviours currently available in Lantz that it would be good to add:
- I would add an extra subclass (`Set`) for values that are only within a valid set. It is like an identity Mapping. It is very useful to check for valid values in the application, before you actually communicate with the instrument.
- I would add a `read_once` modifier to Features. In Lantz we use this to flags properties that do not change and therefore can be cached (i.e. *IDN?)

About the other suff:
- logging: I would not log communication in Lantz. Using PyVISA (and friends) as the single entry point to Message Based communication provides all the communication. We just need to show in the docs how to capture the log or provide a convenience method to redirect the log.
- repeating: I would provide a `retries` attribute (defaulting to 0).
- Error: Not all instruments have an status byte. My suggestions is to do it per class, maybe with a Mixin.
  For example, in Lantz you can override `query` to do the error handling there. It works nicely and I am sure we can extend it.

**Comment by MatthieuDartiailh** [comment]
Bool : The alias was mainly to allow things like 'ON', 'OFF' for output for more readability but I agree this could easily be abused.
Scalar : I am fine with the word limits. Should then the user define a limit getter on the driver ?
Set : For me the set is handled specifying a values keyword which allows to use for Int, Float, etc
ReadOnce : I think that by default all values should be cached and if they are never invalidated we would get the same behavior, wouldn't we ?
I like the logging step being handled somewhere else. I am fine with retry we should however decide where the errors which trigger one should be defined. For the error the way I do it is to rely on a default_check_operation defined on the driver, and called in the post_ methods, which raises an Exception if something went wrong. Would that work for you ?

**Comment by hgrecco** [comment]
About ReadOnce (which I actually called `read_once`, a keyword argument for a Feature). I think I was not clear enough. I am not proposing to have an optional cache. The cache should always be active. I propose to flag those features that will never change so if you read them once, you just need to return the cache value.

Consider the frequency of a signal generator. You need to ask it every time because the user might have changed it manually (if you cannot lock the front panel). But the output of `*IDN?` will never change.

Set: sounds good to me.

Regarding the `default_check_operation`, it depends on the signature. I would classify the instruments in two kinds: 
1. error is reported in the response to the query
2. error is reported in a different way (status byte or another query)

For **1**, the error checker needs to know the response. And the simplest way to do this is to override `query` or make  `default_check_operation` accepts the response as argument.

**Comment by MatthieuDartiailh** [comment]
I must say i don't see the point of having a cache if you don't return the cache value if one exists. If you fear that a user will mess up your instrument state you should deactivate the cache. To me the cache is mostly used when automating operations.

If you are doing a set operation you do not need query you can do a simple write and read later the buffet if the check. I have however no issue in adding an optional arg on which the check is run.

**Comment by hgrecco** [comment]
I think I understand the source of the confusion. We are thinking on different defaults for getters:

If we define:
- current value: the value in the instrument
- cached value: the value in the cache 
- new value: the value given by the user to the setter.

I am proposing the following:
- **get** operations always return the **current** value
- **set** operations are performed only if the new value is different from the cache

Optionally you can:
- return the **cached** value
- set operations can be forced

I think that you are proposing:
- **get** operations always return the **cached** value
- **set** operations are performed only if the new value is different from the cache.

Optionally you can:
- invalidate the cache and return the **current** value
- set operations can be forced

**Comment by MatthieuDartiailh** [comment]
Yes this is actually what I have in mind as when for example checking the mode I wish to avoid useless communications. If we don't do that the checking capabilities of the Feature becomes a bit meaningless as they will be too expensive.
We could have a keyword to mark that a value cannot change so that even when the cache is deactivated those values would be cached nonetheless.

**Comment by MatthieuDartiailh** [comment]
Actually another thing that I think desirable is the automatic locking of instrument access when getting and setting (this avoid issue in multithreaded application and is most useful for multi-channels instruments). This is easy to implement and has virtually no overhead.

**Comment by alexforencich** [comment]
How much of an issue is an extra *idn? when the cache is disabled?  I presume the user script will likely not be trying to query the instrument manufacturer at a high frequency during normal operation, even with the cache disabled.  I would imagine the same thing would apply to most of this sort of read once command.  

On the other hand, one thing that might be interesting is having an option to force all sets - e.g. cache it, but don't check against the cached value when setting, only return the cached value when getting.  This could be one step down from invalidating the entire cache.

**Comment by MatthieuDartiailh** [comment]
I agree with you @alexforencich. Those commands are the ones likely to be queried very frequently. Forcing all set might be an interesting option I will include it in my proposal.

**Comment by alexforencich** [comment]
Maybe a good solution would be to have to settings, one to disable cache in reads and one to disable cache on writes.  Turning cache off on reads would work like @hgrecco is proposing while turning cache off on writes would work like python-ivi does now.  Turning them both on would minimize the number of transactions.  

One other thing that might be interesting is conditional cache clearing.  I think there may be a few properties in python-ivi that do this.  If not, it at least supports it.  With conditional clearing, other properties can be invalidated under arbitrary conditions.  

Another thing that we should figure out how to implement is a 'many to many' mapping system.  This sort of thing is mainly required for cross-driver compatibility, and it can be a bit of a pain to implement.  One example of what I am talking about here is mapping the IVI acquisition type and acquisition sample mode properties to the one acquire:mode command on Agilent oscilloscopes.  In this case, acquisition type can be normal, peak detect, or high resolution while acquisition type can be real time, equivalent time, or segmented.  These then need to be mapped into the acquire:mode options of etime, rtime, pdet, hres, segm, segp, and segh in an intelligent manner.  The method that I have been using is that if you change or query one property, it will figure out the mapping and then update both on the driver side.  For this case it probably would not be terribly complicated to separate out and handle each property separately and just invalidate the cache on the other one, but I don't think this is always possible.  A more complicated example is the scope trigger setting.  In IVI, there are 'width' and 'glitch' trigger types that both map to 'glitch', but with different qualifier settings, (greater, less, or range).  Additionally, the IVI trigger mode 'ac_line' is implemented as 'edge' with the source set to 'line'.  And then different glitch conditions (greater or less) require setting either the low or high threshold on the scope.  Check out agilentBaseScope around line 1100 and 1250 for all of this joy.  This sort of thing is not terribly difficult to write in python-ivi since everything is just getters and setters.  However, this may not be so simple to implement if everything is compartmentalized too much.  

Why does jumping through all of these hoops matter?  Well, if we want instruments to be interchangeable (which I think is a good thing to strive for as much as possible) then we need to write the drivers to have compatible APIs, and this may mean doing quite a bit of remapping to get everything to match up correctly.

**Comment by MatthieuDartiailh** [comment]
I will try to implement the system you propose with two flags. It looks quite reasonable.

We could extend the discard kwarg support to include running a check before clearing the cache. I will first focus on systematic discarding and we can come back to that after.

Your third point is actually tough. I have no magic idea for this one ... at worst we could just write by hand the setter and getter (always possible).

**Comment by alexforencich** [comment]
Well, I think writing the getter and setter directly may be the only way to go, as this sort of mapping could be very complicated and annoying to implement.  However, this actually needs to be possible - is it possible to query other features or set or invalidate the cache of other features from within a getter or setter?  If this is possible, then this should be possible to implement manually on an as-needed basis.

**Comment by MatthieuDartiailh** [comment]
Yes this is definitively possible. The only thing that would prevent it is the lock but as we use a re-entrant lock there is no issue. However we will have to be careful about recursion loops (a getter calling another and back to the first ad infinitum ....)

**Comment by hgrecco** [comment]
Let's forget about `read_once` for a sec and go back to the getter and setter. I will give you more arguments later.

I think the way we use the cache is intimately related to the way we think about properties and methods (Features and Actions) in our drivers. If you always return the cache, then obtaining a value that you do not control is **always** an Action. For example, if you have multimeter and you want to get the voltage

``` python
>>> print(inst.voltage)
```

will not be an option. You will have to do

``` python
>>> print(inst.measure_voltage())
```

(the names are not important).

**Comment by MatthieuDartiailh** [comment]
I would actually favor using action for measurement, to make a clear difference between a fetch (value is read in the buffer of the instrument directly) and a measure (the instrument tries to be smart about the measure and optimize some parameters). The distinction is common in multimeters for example. But I am not sure about others. I though we would discuss this later but miss the link.

However if the cache is an option (per Feature) defaulting to True you could still set False for a measure so that you always read the instrument.

**Comment by hgrecco** [comment]
I do not like the idea of the cache being an option (Again if you thought this was my argument with read_once, it was not. I apologize for the confusion). I rather have either my default or yours but nothing in between because it is confusing for the user. But I do think that this is tightly related to other stuff in the driver that we need to define.

**Comment by MatthieuDartiailh** [comment]
I will open a new issue related to the problem of reading changing values and explain the link to the cache issue.

**Comment by MatthieuDartiailh** [comment]
I tried to move a bit forward with the implementation of the base Feature class. One thing that was going to be an issue was when we had a lot of things to do at a particular stage as determined by the kwargs. Rebuilding a new function each time would be a mess so I a proposing a composition system in https://github.com/LabPy/lantz_core/pull/2. This allows also to decorate specially named method (_post_get, ...) so that they do not override the standard behavior but are rather prepended, appended, or added before or after another operation, this can be quite handy to avoid calling the methods linked to the kwargs by hand. I tried to make it play well with set_feat in the case in which we have some customization using specially named function in some base classes. That part is not pretty but use case for it should be uncommon and strongly discouraged in the docs, I just like the idea that it is possible in some cases.
